pkg/astnormalization/extends_directive.go

```diff
@@ -0,0 +1,56 @@
+package astnormalization
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+)
+
+/*
+@extends directive on an object will make it a type extension
+type User {
+   id: ID!
+   name: String
+}
+     +
+type User @key(fields: "id") @extends {
+   id: ID! @external
+   name: String
+}
+
+will be,
+
+type User {
+   id: ID!
+   name: String
+}
+*/
+type extendsDirectiveVisitor struct {
+	operation *ast.Document
+}
+
+func extendsDirective(walker *astvisitor.Walker) {
+	v := &extendsDirectiveVisitor{}
+	walker.RegisterEnterDocumentVisitor(v)
+	walker.RegisterEnterObjectTypeDefinitionVisitor(v)
+}
+
+func (v *extendsDirectiveVisitor) EnterDocument(operation, _ *ast.Document) {
+	v.operation = operation
+}
+
+func (v *extendsDirectiveVisitor) EnterObjectTypeDefinition(ref int) {
+	if !v.operation.ObjectTypeDefinitions[ref].Directives.HasDirectiveByName(v.operation, "extends") {
+		return
+	}
+	for i := range v.operation.RootNodes {
```

Comment:
Instead of looping over all root notes, you can use the Index:
https://github.com/TykTechnologies/graphql-go-tools/blob/master/pkg/ast/index.go#L63

It will be a lot faster than looping over all RootNodes on every ObjectTypeDefinition

Reply:
in this case, it wouldn't be helpful since we need to modify directly from the document.RootNodes.